src/lib/auth.ts

```diff
@@ -1,43 +1,32 @@
 import logger from './logger'
 import {disableAuth} from './../config'
 import {NextFunction, Request, Response} from 'express'
+import {PermissionLevel} from './auth/types'
 
-// Simple method to ensure that all `req` objects passed in have
-// sufficient access headers to access secured routes. Any route that specifies
-// `secured` will be rejected without these headers.
-export function secured(req: Request, res: Response, next: NextFunction) {
-  if ((req.session && req.isAuthenticated()) || disableAuth) {
-    delete req.session.authBlockedRedirectUrl
-    next()
-    return
-  }
-  logger.info('Unauthenticated request to resource, redirecting to auth')
-  if (req.session) {
-    req.session.authBlockedRedirectUrl = req.originalUrl
-  }
-  res.redirect('/auth')
-}
+export function secured(permissionLevel: PermissionLevel) {
+  return function checkUserIsAuthenticatedAndPermitted(req: Request, res: Response, next: NextFunction) {
+    if (disableAuth) {
+      delete req.session.authBlockedRedirectUrl
+      return next()
+    }
+    if (req.session && req.isAuthenticated()) {
+      if (req.user.permissionLevel && req.user.permissionLevel >= permissionLevel) {
+        delete req.session.authBlockedRedirectUrl
+        return next()
+      }
+      logger.info(`User ${req.user.username} does not have sufficient permissions to access path`)
```

Comment:
It might be nice to log separately for user completely unknown vs. doesn't have a high enough permission level. I haven't looked through the code path carefully that might already be happening!

Reply:
Yeh it's already happening, if the the user is not authenticated it falls through to the code outside the `if` and logs something different

Reply:
```logger.info('Unauthenticated request to resource, redirecting to auth')```